### ISSUE TITLE
Issue with diagnostic settings for azure firewall

### ISSUE THREAD

**Issue by jaspreet-kakkar** [hashicorp/terraform-provider-azurerm#15639]
`terraform v1.0.9`
`azurerm v2.96.0` 

We are deploying azure firewall using the resource azurerm_firewall and adding diagnostic settings using the resource azurerm_monitor_diagnostic_setting pointing to an azure log analytics worksapce. Please see the plan below:

```
# module.firewall.module.logs[0].azurerm_monitor_diagnostic_setting.main will be created
  + resource "azurerm_monitor_diagnostic_setting" "main" {
      + id                             = (known after apply)
      + log_analytics_destination_type = "Dedicated"
      + log_analytics_workspace_id     = "/subscriptions/[subscription-id]/resourceGroups/[rg-name]/providers/Microsoft.OperationalInsights/workspaces/[law-resource-name]"
      + name                           = "hub-vnet-fw-diagnostic-settings"
      + target_resource_id             = "/subscriptions/[subscription-id]/resourceGroups/[rg-name]/providers/Microsoft.Network/azureFirewalls/[firewall-resource-name]"

      + log {
          + category = "AzureFirewallApplicationRule"
          + enabled  = true

          + retention_policy {
              + days    = 7
              + enabled = true
            }
        }
      + log {
          + category = "AzureFirewallDnsProxy"
          + enabled  = true

          + retention_policy {
              + days    = 7
              + enabled = true
            }
        }
      + log {
          + category = "AzureFirewallNetworkRule"
          + enabled  = true

          + retention_policy {
              + days    = 7
              + enabled = true
            }
        }

      + metric {
          + category = "AllMetrics"
          + enabled  = false
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.
```

`tf apply` works fine, no errors. I can see the diagnostic settings applied correctly in the azure portal. However, no logs are ingested in the log analytics workspace. I raised a ticket with MS support. They did a comparison with applying the diagnostic settings through terraform and applying through the portal and they found that the the provider is using a different api-version as compared to the azure portal. This needs to be fixed and the provider should be using the correct api-version.

```
ARM request for portal deployment
2022-02-17 22:21:46.7499165     HttpOutgoingRequestEndWithSuccess PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/MICROSOFT.NETWORK/AZUREFIREWALLS/PROVIDERS/MICROSOFT.INSIGHTS/DIAGNOSTICSETTINGS/                200         PUT        2021-05-01-preview                        MICROSOFT.INSIGHTS   DIAGNOSTICSETTINGS                [https://australiaeast.monitoring.prod.windows.azure.net:443/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall-name]/providers/microsoft.insights/diagnosticSettings/[diagnostic-setting-name]?api-version=2021-05-01-preview](https://australiaeast.monitoring.prod.windows.azure.net/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall-name]/providers/microsoft.insights/diagnosticSettings/[diagnostic-setting-name]?api-version=2021-05-01-preview)
2022-02-17 22:21:47.2451394     HttpIncomingRequestEndWithSuccess PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/MICROSOFT.NETWORK/AZUREFIREWALLS/PROVIDERS/MICROSOFT.INSIGHTS/DIAGNOSTICSETTINGS/                200         PUT        2021-05-01-preview        Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36             MICROSOFT.INSIGHTS   DIAGNOSTICSETTINGS                [https://management.azure.com:443/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall resource name]/providers/microsoft.insights/diagnosticSettings/[diagnostic-setting-name]?api-version=2021-05-01-preview](https://management.azure.com/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall resource name]/providers/microsoft.insights/diagnosticSettings/[diagnostic-setting-name]?api-version=2021-05-01-preview)


ARM request for Terraform deployment
2022-02-22 05:34:54.3036910     HttpIncomingRequestStart PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/MICROSOFT.NETWORK/AZUREFIREWALLS/PROVIDERS/MICROSOFT.INSIGHTS/DIAGNOSTICSETTINGS/                -1            PUT        2017-05-01-preview        Go/go1.17.5 (amd64-linux) go-autorest/v14.2.1 Azure-SDK-For-Go/v61.4.0 insights/2021-07-01-preview HashiCorp Terraform/1.0.9 (+https://www.terraform.io) Terraform Plugin SDK/2.10.1 terraform-provider-azurerm/2.97.0 pid-222c6c49-1b0a-5959-a213-6608f9eb8820        MICROSOFT.INSIGHTS                DIAGNOSTICSETTINGS   [https://management.azure.com:443/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall resource name]/providers/Microsoft.Insights/diagnosticSettings/[diagnostic setting name]?api-version=2017-05-01-preview](https://management.azure.com/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall resource name]/providers/Microsoft.Insights/diagnosticSettings/[diagnostic setting name]?api-version=2017-05-01-preview)
2022-02-22 05:34:58.3725132     HttpIncomingRequestEndWithSuccess PUT/SUBSCRIPTIONS/RESOURCEGROUPS/PROVIDERS/MICROSOFT.NETWORK/AZUREFIREWALLS/PROVIDERS/MICROSOFT.INSIGHTS/DIAGNOSTICSETTINGS/                200         PUT        2017-05-01-preview        Go/go1.17.5 (amd64-linux) go-autorest/v14.2.1 Azure-SDK-For-Go/v61.4.0 insights/2021-07-01-preview HashiCorp Terraform/1.0.9 (+https://www.terraform.io) Terraform Plugin SDK/2.10.1 terraform-provider-azurerm/2.97.0 pid-222c6c49-1b0a-5959-a213-6608f9eb8820        MICROSOFT.INSIGHTS                DIAGNOSTICSETTINGS   [https://management.azure.com:443/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall resource name]/providers/Microsoft.Insights/diagnosticSettings/[diagnostic setting name]?api-version=2017-05-01-preview](https://management.azure.com/subscriptions/[subscription-id]/resourceGroups/[resource-group-name]/providers/Microsoft.Network/azureFirewalls/[firewall resource name]/providers/Microsoft.Insights/diagnosticSettings/[diagnostic setting name]?api-version=2017-05-01-preview)
```

**Comment by tombuildsstuff** [comment]
hi @jaspreet-kakkar 

Thanks for opening this issue.

At the time of writing we're using the Azure SDK package for `2021-07-01-preview` which is published by the Service Team using the API Versions they recommend for this particular grouping of services. Unfortunately the API versions which may comprise an SDK version aren't 1:1 - and whilst this is something we're looking to resolve, ultimately this would need to be fixed by the Service Team since this is coming from the Azure SDK (which is in turn coming from the composite Swagger definition that the Service Team is publishing).

As such unfortunately this is something which'd need to be fixed by the Service Team rather than something we can 
As such unfortunately this needs to be resolved by the Service Team and as such I'd recommend [opening an issue on the Rest API Specs repository](https://github.com/Azure/azure-rest-api-specs) where someone from the Service Team should be able to look into this - however since this needs to be fixed by the Service Team rather than something we can currently control in Terraform, I'm going to close this issue for the moment.

Thanks!

**Comment by jaspreet-kakkar** [comment]
Update: I found out today that the cause of my issue wasn't the that the api-version used by the terraform azurerm provider. In fact,  I found out that in our terraform configuration, we were passing a parameter which specifies to create resource-specific log tables as compared to doing it in azure portal which configures the log analytics workspace in AzureDiagnostics table mode which is legacy mode. I changed that terraform parameter to AzureDiagnostics mode and it seems to be working fine now.
